File: examples/scala/src/main/scala/com/xebia/functional/xef/scala/conversation/fields/Book.scala
package com.xebia.functional.xef.scala.auto.fields

import com.xebia.functional.xef.prompt.Prompt
import com.xebia.functional.xef.scala.auto.*
import io.circe.Decoder

@Description("A book")
case class Book(
    @Description("the name of the book") name: String,
    @Description("the author of the book") author: String,
    @Description("A 50 word paragraph with a summary of this book") summary: String
) derives SerialDescriptor,
      Decoder

def summarizeBook(title: String, author: String)(using conversation: ScalaConversation): Book =
  prompt(Prompt(s"$title by $author summary."))

@main def runBook: Unit =
  conversation {
    val toKillAMockingBird = summarizeBook("To Kill a Mockingbird", "Harper Lee")
    println(s"${toKillAMockingBird.name} by ${toKillAMockingBird.author} summary:\n ${toKillAMockingBird.summary}")
  }

File: examples/src/main/kotlin/com/xebia/functional/xef/conversation/fields/Book.kt
package com.xebia.functional.xef.conversation.fields

import ai.xef.openai.StandardModel
import com.xebia.functional.openai.models.CreateChatCompletionRequestModel
import com.xebia.functional.xef.conversation.Conversation
import com.xebia.functional.xef.conversation.Description
import com.xebia.functional.xef.prompt.Prompt
import kotlinx.serialization.Serializable

@Serializable
data class Book(
  @Description("The title of the book.") val title: String,
  @Description("The author of the book.") val author: String,
  @Description("A 50 word summary of the book.") val summary: String
)

suspend fun main() {
  Conversation {
    val model = StandardModel(CreateChatCompletionRequestModel.gpt_3_5_turbo_16k_0613)
    val toKillAMockingbird: Book = prompt(Prompt(model, "To Kill a Mockingbird by Harper Lee"))
    println(toKillAMockingbird)
  }
}
